### PR TITLE
[SPARK-49183][SQL] V2SessionCatalog.createTable should respect PROP_IS_MANAGED_LOCATION

### PR DESCRIPTION
**Pull request by cloud-fan**
### What changes were proposed in this pull request?

Even if the table definition has a location, the table should be a managed table if `PROP_IS_MANAGED_LOCATION` is specified, in `V2SessionCatalog.createTable`.

### Why are the changes needed?

It's a bug fix. A custom `spark_catalog` may generate custom location for managed table and delegate the actual table creation to `V2SessionCatalog`. The table should still be a managed table if `PROP_IS_MANAGED_LOCATION` is specified.

### Does this PR introduce _any_ user-facing change?

Yes, now users who use custom `spark_catalog` that generates custom location for managed table, can correctly create managed tables.

### How was this patch tested?

a new test

### Was this patch authored or co-authored using generative AI tooling?

no

### CODE REVIEW

**Comment by cloud-fan** [comment]
@yaooqinn @amaliujia

**Comment by yaooqinn** [comment]
Do we need to update the API doc to make it more precise? 

```
 /**
   * A reserved property to indicate that the table location is managed, not user-specified.
   * If this property is "true", SHOW CREATE TABLE will not generate the LOCATION clause.
   */
```

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/V2SessionCatalog.scala [hunk 1]
```diff
@@ -170,7 +170,8 @@ class V2SessionCatalog(catalog: SessionCatalog)
     val storage = DataSource.buildStorageFormatFromOptions(toOptions(tableProperties))
       .copy(locationUri = location.map(CatalogUtils.stringToURI))
     val isExternal = properties.containsKey(TableCatalog.PROP_EXTERNAL)
-    val tableType = if (isExternal || location.isDefined) {
+    val isManagedLocation = properties.containsKey(TableCatalog.PROP_IS_MANAGED_LOCATION)
```
Comment (yaooqinn) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/V2SessionCatalog.scala:173]: do we need to test if it is true or not?

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/V2SessionCatalog.scala:173]: I followed how `PROP_EXTERNAL` is handled.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/ShowCreateTableExec.scala [hunk 2]
```diff
@@ -120,7 +120,7 @@ case class ShowCreateTableExec(
   private def showTableLocation(table: Table, builder: StringBuilder): Unit = {
     val isManagedOption = Option(table.properties.get(TableCatalog.PROP_IS_MANAGED_LOCATION))
     // Only generate LOCATION clause if it's not managed.
-    if (isManagedOption.forall(_.equalsIgnoreCase("false"))) {
+    if (isManagedOption.isEmpty || !isManagedOption.get.equalsIgnoreCase("true")) {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/ShowCreateTableExec.scala:123]: A small fix to match the document. `true` means effective, non `true` means not effective.

**Review by yaooqinn — APPROVED** [review]

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/TableCatalog.java [hunk 3]
```diff
@@ -52,7 +52,8 @@ public interface TableCatalog extends CatalogPlugin {
 
   /**
    * A reserved property to indicate that the table location is managed, not user-specified.
-   * If this property is "true", SHOW CREATE TABLE will not generate the LOCATION clause.
+   * If this property is "true", it means it's a managed table even if it has a location. As an
+   * example, SHOW CREATE TABLE will not generate the LOCATION clause.
```
Comment (amaliujia) [inline sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/TableCatalog.java:56]: Do we need to also document non true case? 

Is non true value invalid, or it is valid but ignored, etc.?

File: sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2SQLSuite.scala [hunk 4]
```diff
@@ -3612,6 +3616,20 @@ class DataSourceV2SQLSuiteV1Filter
...
+  test("SPARK-49183: custom spark_catalog generates location for managed tables") {
+    // Reset CatalogManager to clear the materialized `spark_catalog` instance, so that we can
+    // configure a new implementation.
+    spark.sessionState.catalogManager.reset()
+    withSQLConf(V2_SESSION_CATALOG_IMPLEMENTATION.key -> classOf[SimpleDelegatingCatalog].getName) {
+      withTable("t") {
+        sql(s"CREATE TABLE t (i INT) USING $v2Format")
+        val table = catalog(SESSION_CATALOG_NAME).asTableCatalog
+          .loadTable(Identifier.of(Array("default"), "t"))
+        assert(!table.properties().containsKey(TableCatalog.PROP_EXTERNAL))
```
Comment (amaliujia) [inline sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2SQLSuite.scala:3628]: do we need to also test `SHOW CREATE TABLE` or `DESC TABLE` that this is a managed table without a location?

**Review by dongjoon-hyun — APPROVED** [review]

**Review by HyukjinKwon — APPROVED** [review]

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/TableCatalog.java (hunk 3 above)
Comment (cloud-fan) [reply sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/TableCatalog.java:56]: The non-true case is hard to document, as treating tables with location as external tables is a Hive-specific behavior.

File: sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2SQLSuite.scala (hunk 4 above)
Comment (cloud-fan) [reply sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2SQLSuite.scala:3628]: v1 table property is already the source of truth.

**Review by amaliujia — APPROVED** [review]
LGTM

**Comment by cloud-fan** [comment]
thanks for the review, merging to master/3.5!

**Comment by ulysses-you** [comment]
it seems the branch-3.5 CI failed..
```
[error] /home/runner/work/spark/spark/sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2SQLSuite.scala:3401:24: type Map is not a member of package org.apache.spark.sql.connector.util
[error]       properties: util.Map[String, String]): Table = {
[error]                        ^
[error] /home/runner/work/spark/spark/sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2SQLSuite.scala:3402:29: type HashMap is not a member of package org.apache.spark.sql.connector.util
[error]     val newProps = new util.HashMap[String, String]
[error]                             ^
```

**Pull request closed without merging**